File: app/src/main/java/com/donald/musictheoryapp/Music/ScoreView/SNoteArrow.java
package com.donald.musictheoryapp.Music.ScoreView;

import android.graphics.Canvas;

public class SNoteArrow extends ScoreStamp
{
	private String label;
	private final char glyph;

	SNoteArrow(ScoreView view)
	{
		super(view);
		glyph = U.ARROW_BLACK_DOWN;
	}

	protected void setLabel(String label)
	{
		this.label = label;
	}

	@Override
	protected void onDraw(Canvas canvas, float posX, float posY)
	{
		float glyphSize = glyphSize() * 1.2f;
		float vOffsetOnStaff = -staffStepHeight() * 1 - staffHeight();
		float hOffsetOnStaff = -noteWidth() / 2;

		posX += hOffsetOnStaff;
		posY += vOffsetOnStaff;

		float defaultGlyphSize = glyphPaint().getTextSize();
		glyphPaint().setTextSize(glyphSize);
		canvas.drawText(String.valueOf(glyph),
			posX,
			posY,
			glyphPaint());
		glyphPaint().setTextSize(defaultGlyphSize);

		if(label != null)
		{
			canvas.drawText(label,
				posX - noteWidth() * 1.5f,
				posY - staffStepHeight() * 2,
				textPaint());
		}
	}
}

File: app/src/main/java/com/donald/musictheoryapp/music/scoreview/SNoteArrow.kt
package com.donald.musictheoryapp.music.scoreview

import android.graphics.Canvas
import com.donald.musictheoryapp.music.MusicXML.Note

internal class SNoteArrow(view: ScoreView) : ScoreStamp(view) {

    var note: Note? = null
    private var draws: Boolean = false
    private var label: String = ""
    private val glyph: Char = U.ARROW_BLACK_DOWN
    override val relContentWidth: Float // IMPLEMENT
        get() = if (draws) {
            REL_NOTE_ARROW_ARROW_WIDTH + REL_NOTE_ARROW_SPACING + REL_NOTE_ARROW_LABEL_WIDTH + REL_MARGIN_AFTER_ARROW
        } else {
            0F
        }

    fun set(note: Note) {
        val noteArrow = note.notations?.noteArrow
        if (noteArrow != null) {
            draws = true
            label = noteArrow.label
        } else {
            draws = false
            label = ""
        }
    }

    override fun onDraw(canvas: Canvas, posX: Float, posY: Float) {
        //val note = note ?: return
        //set(note)
        if (draws) {
            var currentPosX = posX

            val textSize = textPaint.textSize * 0.8F
            val defaultTextSize = textPaint.textSize
            textPaint.textSize = textSize
            canvas.drawText(
                label,
                currentPosX,
                posY - staffStepHeight * 3,
                textPaint
            )
            textPaint.textSize = defaultTextSize

            currentPosX += (REL_NOTE_ARROW_LABEL_WIDTH + REL_NOTE_ARROW_SPACING) * glyphSize

            canvas.drawText(
                glyph.toString(),
                currentPosX,
                posY - staffStepHeight,
                glyphPaint
            )
        }
    }

}